### PR TITLE
ENH: #10143 Function to walk the group hierarchy of a PyTables HDF5 file

### LINKED ISSUE #10143: ENH: Iterate over HDF store hierarchically
I am relatively new to pandas, but have been using raw HDF5 files for a long itme, so apologies if this feature or something equivalent already exists.

As stated in the documentation, it is possible to create hierarchies of data stores:

``` python
store.put('/first_group/df1', df1)
store.put('/first_group/df2', df2)
store.put('/second_group/df3', df3)
```

However, it seems that `store.keys()` can only provide a flattened list of all available groups.  It would be nice to be able to walk the HDF5 file hierarchically.  A couple examples (this is borrowing a bit from the h5py API):

``` python
# do stuff on each subgroup/substore in /first_group
for s in store['first_group'].values() :
    # do stuff on df1 and df2

# visit each group at the root level
for s in store.values() :
    if 'df1' in s :
        # do stuff on any store named 'df1'
```

The main use case I would have for something like this is that I have many dataframes with different schema but would like to group some of them together and then iterate through those groups. 

Best way I can think to emulate this is to parse the strings in `store.keys()`, but that is pretty ugly.

### PR DESCRIPTION
**Pull request by stephenpascoe**
closes #10143 

This implementation is inspired by os.walk and follows the interface as much as possible.

### CODE REVIEW

**Comment by jreback** [comment]
any reason not to call this just walk?

**Comment by jreback** [comment]
pls add a note in whatsnew/0.17.0 in enhancements

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -1038,6 +1038,38 @@ def groups(self):
...
+                continue
+
+            groups = []
+            frames = []
+            for child in g._v_children.values():
+                pandas_type = getattr(child._v_attrs, 'pandas_type', None)
+                if (getattr(child._v_attrs, 'CLASS', None) == 'GROUP'
+                    and pandas_type is None):
+                    groups.append(child._v_name)
+                elif pandas_type == 'frame':
```
Comment (jreback) [inline pandas/io/pytables.py]: yield ANY pandas_type objects, so don't need this complicated if

File: pandas/io/pytables.py [hunk 2]
```diff
@@ -1038,6 +1038,38 @@ def groups(self):
...
+        for g in self._handle.walk_groups():
+            if (getattr(g, '_v_name', None) is None
+                or getattr(g._v_attrs, 'pandas_type', None) == 'frame'):
+                continue
+
+            groups = []
+            frames = []
+            for child in g._v_children.values():
+                pandas_type = getattr(child._v_attrs, 'pandas_type', None)
+                if (getattr(child._v_attrs, 'CLASS', None) == 'GROUP'
```
Comment (jreback) [inline pandas/io/pytables.py]: yield as you walk, not at the end

File: pandas/io/tests/test_pytables.py [hunk 3]
```diff
@@ -4813,6 +4813,41 @@ def test_read_nokey(self):
...
+    # GH10143
+    def test_walk_groups(self):
+        with tm.ensure_clean('walk_groups.hdf') as filename:
+            store = HDFStore(filename, 'w')
+
+            dfs = {
+                'df1': pd.DataFrame([1,2,3]),
+                'df2': pd.DataFrame([4,5,6]),
+                'df3': pd.DataFrame([6,7,8]),
+                'df4': pd.DataFrame([9,10,11]),
```
Comment (jreback) [inline pandas/io/tests/test_pytables.py]: add a Series and a PyTables object here (e.g. non-pandas) as well

**Comment by stephenpascoe** [comment]
I've renamed walk() and included Series objects.  However, I think you have a different API in mind.  I am deliberately not yielding each Pandas object individually but yielding each PyTables group name, along with a tuple of its contents.  This follows the os.walk API.  I.e. each yield is 

  `(group_path, [subgroup_name, ...], [subobj_name, ...])`

I think there are several advantages:

1. The consumer can see the difference between groups and Pandas objects
2. Future extension could allow pruning of the search space by mutating the yielded lists, as is possible with os.walk.

Note also:
* Some testing of node type is necessary during walk because a Pandas object is also a group to PyTables.
* All non-pandas leaves are ignored.  walk() will only yield groups and Pandas objects.

Please let me know what you think before I write something in whatsnew/0.17.0

File: pandas/io/pytables.py [hunk 4]
```diff
@@ -1038,6 +1038,37 @@ def groups(self):
                  g._v_name != u('table')))
         ]
 
+    def walk(self):
+        """ Walk the pytables group hierarchy yielding the group name and pandas object names
+        for each group.  Any non-pandas PyTables objects that are not a group will be ignored.
+
+        Returns
+        -------
+        A generator yielding tuples (`path`, `groups`, `frames`) where:
```
Comment (jreback) [inline pandas/io/pytables.py]: frames -> leaves

File: pandas/io/pytables.py [hunk 5]
```diff
@@ -1038,6 +1038,37 @@ def groups(self):
...
+
+        - `path` is the full path to a group,
+        - `groups` is a list of group names contained in `path`
+        - `leaves` is a list of pandas object names contained in `path`
+
+        """
+        _tables()
+        self._check_if_open()
+        for g in self._handle.walk_groups():
+            if getattr(g._v_attrs, 'pandas_type', None) is not None:
```
Comment (jreback) [inline pandas/io/pytables.py]: shouldn't this be ``is None``?

File: pandas/io/tests/test_pytables.py [hunk 6]
```diff
@@ -4813,6 +4813,47 @@ def test_read_nokey(self):
...
+        objs = {
+            'df1': pd.DataFrame([1,2,3]),
+            'df2': pd.DataFrame([4,5,6]),
+            'df3': pd.DataFrame([6,7,8]),
+            'df4': pd.DataFrame([9,10,11]),
+            's1': pd.Series([10,9,8]),
+            'a1': np.array([[1,2,3], [4,5,6]])
+            }
+
+        with tm.ensure_clean('walk_groups.hdf') as filename:
```
Comment (jreback) [inline pandas/io/tests/test_pytables.py]: you can just use ``ensure_clean_store``

**Comment by jreback** [comment]
ok, looks reasonable. Pls add a whatsnew (and add this to the docs in io.rst), separate-sub-section in the HDF5 section.

File: pandas/io/pytables.py (hunk 5 above)
Comment (stephenpascoe) [inline pandas/io/pytables.py]: No, HDF5 groups that have 'pandas_type' attribute will be group wrappers around dataframe/series objects.  Every Pandas object is wrapped in an HDF5 group.

Comment (jreback) [inline pandas/io/pytables.py]: ahh ok

**Comment by jreback** [comment]
closing, pls reopen if you can fix according to comments

**Pull request closed without merging**